### DISCUSSION TITLE
LDAP: Do not import users when searching

### DISCUSSION THREAD

**Discussion by fjf2002** [keycloak/keycloak#24140] [Q&A]
I have a LDAP User federation configured, using Kerberos.

As far as I understand the plethora of LDAP settings and the corresponding keycloak behaviour, I cannot have all three of the following conditions:

* Ability to add groups to the user on the keycloak-side (i. e. Edit-mode=UNSYNCED)
* Import the LDAP user into the keycloak DB on login (Import users = on ?)
* Do not auto import users from LDAP when searching for users in the keycloak admin backend(?)

How disable the latter?

Thanks for your advice or your enhancement regarding this request.

### Motivation

I simply do not want to import personal data from the LDAP that I clearly do not want and need, also having in mind the GDPR.

### Details

Suggestion
* Clearly state in the manual 
  * what the "import users" setting does,
  * that even a search triggers user importing,
* Offer a setting to turn the latter off(?)

**Comment by fjf2002** [comment]
**bump** any advances on this topic?

**Reply by fjf2002** [reply]
**bump** does anybody want to contribute to this topic?

**Reply by darius-m** [reply]
I am not entirely familiar with all the settings, so I might be wrong, but from my understanding user synchronization is a feature that is meant to synchronize users in the LDAP database to Keycloak, and possibly update user attributes if mappers are not set to always read the data from LDAP.

Even if user synchronization is disabled, some user data (e.g., username or email) is still required for all Keycloak users who log in, as you must be able to correlate user sessions or save some user-specific data like stored 2FA tokens or client consents with the user.

Searching for users also probably requires creating some entries, since some features of the admin portal, like pagination, must be able to display or otherwise handle a subset of users. If this information is not persistent, you would not be able to perform operations reliably. For example, if when searching only some users would be extracted from the database, while others would be constructed dynamically from the LDAP query, you would have to reliably reconstruct the user list every time you click on the "next" or "previous" buttons. You would also have to keep track of each user type in the admin interface, since some users would not have some attributes (e.g., the user ID in the Keycloak database), so performing (bulk) operations on them would not be possible. From my extremely limited testing, it seems that entries are created only for users who appear in the search results, so not all users are synchronized.

That being said, at least in Keycloak 23, it seems that the format for the user ID is different if user synchronization is disabled (it contains `f:ldap-uuid:username`), at least if the user is created through the search process. Users still appear to be created "at some point" in time, and that time does not change, but the user information is not saved in the `USER_ATTRIBUTES` table (it might be in Infinispan, however).

**Comment by fjf2002** [comment]
@darius-m: Thanks for your reply.

> Even if user synchronization is disabled, some user data (e.g., username or email) is still required for all Keycloak users who log in [...]

Yes, of course. But just for **all users who log in**, not more. But I experienced: Keycloak downloads the whole LDAP OU, particularly users that never log into Keycloak. That download seems to happen block-wise. After a few minutes, I noticed some hundred users downloaded from LDAP (that DID NOT log in), after hours even more.

Firstly, I was thinking of some GDPR (or whatever) rule "privacy by design", which I interpreted here as "do not download users from LDAP that aren't neccessary because they do not log in". You know, IF an LDAP user logs in, I consider the login as some kind of consent that the user is willing to store his data into Keycloak; but not I do not have that "consent" beforehand.


Secondly, as you might know, Keycloak distinguishes between "Identity Providers" and "User federation".

- "Identity Provider" Keycloak behaviour gives us: Keycloak users are created on-the-fly as users log into the foreign identity provider (but not before!)
- "User federation" Keycloak behaviour gives us, as I described above: Keycloak users get created from all LDAP users, even if some never log into Keycloak.
- What I am looking for is simply: "Identity Provider Keycloak behaviour" for LDAP. Shouldn't be that difficult, I thought...

**Reply by darius-m** [reply]
I think that what you request does make some sense, but from what I could see there are some fundamental design differences between the two. My understanding of the internals is very limited, so take the information with quite a bit of salt.

An identity provider is simply a different (foreign) SSO that Keycloak redirects the user to when authenticating and trusts that the provided response is valid. The identity provider also offers a compliant implementation of SSO protocols. However, this is basically the only operation you can do - I don't think you can search for users in the Google (for example) user database, even if you have configured federation between the two, nor is it possible to store any users in the remote identity provider from Keycloak.

Federation, on the other hand, is considered more of a (user) database. Neither LDAP, nor AD, offer a full implementation of SSO protocols. Instead they were (actually) built as more general purpose databases - aside from (web) users and passwords you can also store other information like UNIX user and group IDs, home directory paths, policies, DNS entries, and so on. Keycloak uses LDAP as a more traditional storage database, and allows you to dynamically store and fetch users from this database and search for them even if they were not imported in Keycloak's database yet.

**Reply by fjf2002** [reply]
@darius-m: Thanks for your reply, I agree with your remarks.

Nevertheless, is anybody interested in adding my described functionality to keycloak?

### It could simply be:
Keep LDAP as "User federation", but allow simply to disable downloading all users from LDAP. Just download the ones that actually log in.

### A more elaborate way: 
Add LDAP/Kerberos to the supported IDP types (OAuth, SAML, and now additionally LDAP/Kerberos).
